sql-plugin/src/main/311until340-all/scala/com/nvidia/spark/rapids/shims/CastingConfigShim.scala

```diff
@@ -18,11 +18,22 @@ package com.nvidia.spark.rapids.shims
 
 import org.apache.spark.sql.catalyst.expressions.{CastBase, Expression}
 
-object CastIgnoreTimeZoneShim {
+object CastingConfigShim {
   /** Remove TimeZoneId for Cast if needsTimeZone return false. */
   def ignoreTimeZone(e: Expression): Expression = e match {
     case c: CastBase if c.timeZoneId.nonEmpty && !c.needsTimeZone =>
       c.withTimeZone(null)
     case _ => e
   }
+
+  def ansiEnabled(e: Expression): Boolean = {
+      // prior to Spark 3.4.0 we could use CastBase as argument type, but starting 3.4.0 the type is
+      // the case class Cast.
+      // prior to Spark 3.3.0 we could use toString to see if the name of
+      // the cast was "cast" or "ansi_cast" but now the name is always "cast"
+      // so we need to use reflection to access the protected field "ansiEnabled"
+      val m = e.getClass.getDeclaredField("ansiEnabled")
```

Comment:
I am not a fan of using reflection unless we have no other choice. The reflection code was originally only for tests, but in order to shim it the code has been moved into non-test code.  At a minimum we need to have lots of documentation on the method an preferably in the name of the method that this is only to be used for tests and needs to be called with some kind of a cast expression. That or we make it so we don't need to use reflection to know what kind of a cast is happening and the code is good enough to be used in production.

Reply:
I agree, though I don't love the other options. Because `ansiEnabled` is private, ability to retrieve that value is basically limited to reflection. Because the inconsistencies with the `Cast` object, 31x will fail if it isn't shimmed and I've been unable to get reflection to work for the public `ansiEnabled` field.

This is my suggested solution, though I don't love it's cleanliness. `publicAnsiEnabled` will return false for 311 until 340, and 340+ will return the `c.asInstanceOf[Cast].ansiEnabled`. Is this more more acceptable solution?
```
    def isAnsiCast(c: Expression): Boolean = {
      if (cmpSparkVersion(3, 4, 0) < 0) {
        val m = c.getClass.getDeclaredField("ansiEnabled")
        m.setAccessible(true)
        m.getBoolean(c)
      } else {
        CastingConfigShim.publicAnsiEnabled(c)
      }
    }
    ```